### ISSUE TITLE
catfile option not allowed in XQuery library module

### ISSUE THREAD

**Issue by gimsieke** [BaseXdb/basex#1793]
I was finally trying to reap the benefits of #1719 by invoking an XSLT transformation that imports a stylesheet through a catalog-resolved URI. The XSLT transformation is invoked from a library module that declares several functions with RESTXQ annotations. 

BaseX complains: 

```
[basex:options] Database option not allowed in library module: catfile.
```
Is this error a remnant from the situation when catalog files were exclusively used when creating databases, not when, for example, invoking an XSLT transformation from a function in a library module?

**Comment by ChristianGruen** [comment]
The declaration of [BaseX database options](http://docs.basex.org/wiki/XQuery_Extensions#Option_Declarations) is restricted to XQuery main modules. The reason is that library modules can be imported from other modules, and this may result in conflicting values for the same options.

Instead, you can…
1. use a [database pragma](http://docs.basex.org/wiki/XQuery_Extensions#Database_Pragmas) to declare your option locally, or
2. define your RESTXQ functions in a main modules.

Option 1 is what we usually recommend.

**Comment by gimsieke** [comment]
I tried using the pragma approach, but encountered at least two issues with it.

This simplified function declaration works:

```xquery
declare
  %rest:GET
  %rest:path("/chpd/{$work}/mono-hobots2/{$name}")
  %output:method("xml")
function chpd:mono-hobots2(
  $name   as xs:string,
  $work   as xs:string
) as item()* {
  $chpd:xml-header,
  (# db:catfile '../transpect/calabash/xmlcatalog/catalog.xml' #) { 
    <doc>{db:open('CHPD_Main_Monos')[1]}</doc>                      
  }
};
```

As expected, it retrieves a document from the specified DB and wraps it in a `doc` element.

First, I don’t want to wrap the document in a `doc` element. So when I change the inner part to

```xquery
  (# db:catfile '../transpect/calabash/xmlcatalog/catalog.xml' #) { 
    {db:open('CHPD_Main_Monos')[1]}                      
  }
```

I get: `XQST0079] Expecting pragma expression.`

So for the time being I leave the wrapper around the result. What I originally set out to do was to transform the document that was retrieved from the DB. So I changed the expression to:

```xquery
  (# db:catfile '../transpect/calabash/xmlcatalog/catalog.xml' #) { 
    <doc>{xslt:transform(db:open('CHPD_Main_Monos')[1], 
          '../../../transpect/a9s/common/hub2hobots/hub2hobots.xsl')}</doc>                      
  }
```

and then I get the error: `[FODC0002] "" (Line 1): Premature end of file.`

If I omit the pragma like this:

```xquery
    <doc>{xslt:transform(db:open('CHPD_Main_Monos')[1], 
          '../../../transpect/a9s/common/hub2hobots/hub2hobots.xsl')}</doc>                      
```

I get: `[xslt:error] Static error at html on line 1 column 107 of http://transpect.github.io/:
  XTSE0150: The supplied file does not appear to be a stylesheet`
The XSLT contains `<xsl:import href="http://transpect.io/hub2bits/xsl/hub2bits.xsl"/>`. Without the catalog, it will just retrieve an HTML page at http://transpect.io/, therefore the message is totally expected.

So why can’t I use `xslt:transform()` within the pragma block, and why do I have to create a wrapper element?

I can create a full repro if you want.

**Comment by ChristianGruen** [comment]
> I get: `XQST0079] Expecting pragma expression.`

I would guess this is due to the double curly braces. Could you try this?

```xquery
  (# db:catfile '../transpect/calabash/xmlcatalog/catalog.xml' #) { 
    db:open('CHPD_Main_Monos')[1]
  }
```

> I can create a full repro if you want.

Always appreciated. It shouldn't make a difference if you use RESTXQ or not, so a simple query eypressions and attached XML and catalog files should be sufficient.

**Comment by gimsieke** [comment]
> I would guess this is due to the double curly braces. Could you try this?
> 
> ```xq
>   (# db:catfile '../transpect/calabash/xmlcatalog/catalog.xml' #) { 
>     db:open('CHPD_Main_Monos')[1]
>   }
> ```

Yes, this works. But if I use 

```xquery
  (# db:catfile '../transpect/calabash/xmlcatalog/catalog.xml' #) { 
    xslt:transform(db:open('CHPD_Main_Monos')[1], 
          '../../../transpect/a9s/common/hub2hobots/hub2hobots.xsl')
  }    
```

 I get `[FODC0002] "" (Line 1): Premature end of file.`

> > I can create a full repro if you want.
> 
> Always appreciated. It shouldn't make a difference if you use RESTXQ or not, so a simple query eypressions and attached XML and catalog files should be sufficient.

I created [issue1793.zip](https://github.com/BaseXdb/basex/files/4114000/issue1793.zip). If you unzip it in the current working directory and open `issue1793/issue1793_global-catfile-relative.xq` in the GUI, it should transform `doc.xml` using `issue1793.xsl`, which imports `importme.xsl` if the catalog is used, which it should be. (A prerequisite is probably that Saxon is on the classpath, since the XSLT is version 2.0.)

If I execute `issue1793/issue1793_pragma-catfile-relative.xq` afterwards, it will seem to work, too, but only until I restart the GUI. After a restart, I get the error

```
[xslt:error] Error on line 3 column 62 of issue1793.xsl:
  XTSE0165: I/O error reported by XML parser processing
  http://canonical.location/importme.xsl: canonical.location
```
The same error occurs if I use an absolute path or file URI in the pragma variant. So `declare option` seems to initialize `db:catfile` (also for other tabs in the GUI) while the pragma never sets the option.

**Comment by ChristianGruen** [comment]
Hi Gerrit, I found the (non-obvious) cause of confusion: [XQuery pragma contents](https://www.w3.org/TR/xquery-31/) are plain character sequences, they are not interpreted as XQuery literals. You should be fine if you omit the quotes:

```xquery
(# db:catfile issue1793/catalog.xml #) {
  xslt:transform('doc.xml', 'issue1793.xsl')
}
```

**Comment by gimsieke** [comment]
Amazing. I was sure that I already tried it without the quotes. But maybe with the wrong relative path. Now I’m going to see whether I can make it work in the RESTXQ scenario, too. And if everything works, I can update the pragma and catalog documentation in the wiki.

**Comment by gimsieke** [comment]
I added documentation on the [Catalog Resolver](http://docs.basex.org/wiki/Catalog_Resolver) and [Pragma](http://docs.basex.org/wiki/XQuery_Extensions#Database_Pragmas) pages.

Also I found out why the XSLT transformation ended with `[FODC0002] "" (Line 1): Premature end of file.`: The stylesheet was supposed to be invoked in an initial mode. The `#default` mode only produced text nodes, which led to the error. 

Sorry for all the noise, but I think documenting how to use catalogs for XSLT transformations and how to correctly use pragmas for pointing to catalogs was finally worth it.
